### PR TITLE
Fix `mro` of generic subclass 

### LINKED ISSUE #10039: Nested generic inheritance
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

Is there a way for `MyClassSubBool` to be recognized as an instance of `MyClass[bool]` , apart from inheriting the functionality from `MyClassSub`?

If I use multiple inheritance in `class MyClassSubBool(MyClassSub, MyClass[bool])` I do get the example to work, but it doesn't seem the most pythonic way.

Thanks for your help.

### Example Code

```Python
from typing import Generic, TypeVar

from pydantic import BaseModel

T = TypeVar('T')


class MyClass(BaseModel, Generic[T]):
    pass


class MyClassSub(MyClass[T]):
    pass


class MyClassSubBool(MyClassSub[bool]):
    pass


class Model(BaseModel):
    input_bool: MyClass[bool]


if __name__ == '__main__':
    instance_fails = Model(input_bool=MyClassSubBool())
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.7.3
        pydantic-core version: 2.18.4
          pydantic-core build: profile=release pgo=true
               python version: 3.11.4 (main, Sep 12 2023, 11:07:53) [GCC 11.4.0]
                     platform: Linux-6.5.0-44-generic-x86_64-with-glibc2.35
             related packages: typing_extensions-4.10.0
                       commit: unknown
```

### PR DESCRIPTION
**Pull request by kc0506**
## Change Summary

Override `ModelMetaclass.mro` such that indexed generic type can be inserted into `__mro__` of its subclasses.

For example:
```py
class A(BaseModel, Generic[T]): ...
class B(A[T]): ...
class C(B[bool]): ...

# before
assert C.__mro__ == (C, B[bool], B, A, BaseModel, Generic, object)
# after
assert C.__mro__ == (C, B[bool], B, A[bool], A, BaseModel, Generic, object)
```
This allow `C()` to be assigned to `A[bool]`.

I have checked all testcases in `test_generic.py`. This change doesn't break existing tests. Also, I collected all the tests where `__mro__` is affected by this change, and add them into new tests. 

## Related issue number
fix #10039 

## Checklist

* [x] The pull request title is a good summary of the changes - it will be used in the changelog
* [x] Unit tests for the changes exist
* [x] Tests pass on CI
* [x] Documentation reflects the changes where applicable
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

### CODE REVIEW

**Comment by sydney-runkle** [comment]
Hmm, I think the way to solve this is actually going to be revalidating instances of classes that have a shared generic parent. Happy to think about this proposal more tomorrow.

@dmontagu, I haven't looked in depth, but this is up your alley, and might be worth a glance.

**Review by dmontagu — CHANGES_REQUESTED** [review]
So I think this may make sense in principle, but if we merge it I think we should try to do a better job of substituting the parameters in intermediate classes even when they aren’t the same.

E.g., if I have `class A(BaseModel, Generic[T]):…` and `class B(A[T], Generic[S]): …` we should have `A[int]` in the mro of `B[int, str]`. I don’t think it should be too hard to achieve this (or at least do a better job than exact equality checking), but if it is unreasonably hard to follow through the substitutions we don’t need to let perfect get in the way of better.

That said, I am not 100% sure if this is better than revalidating into the other generic type. But I think this will result in better behavior in the cases that are affected, so I’m inclined to merge this approach.

**Comment by kc0506** [comment]
@dmontagu Thanks for the review!
I made some changes, so now this can work:
```py
class A(BaseModel, Generic[T1]): ...
class B1(A[T1], Generic[T1, T2]): ...

assert A[int] in B1[T1, str][int].__mro__
assert A[int] in B1[int, str].__mro__
```

File: pydantic/_internal/_model_construction.py [hunk 1]
```diff
@@ -241,6 +243,61 @@ def __getattr__(self, item: str) -> Any:
...
+            target_params: tuple[TypeVar] = origin.__pydantic_generic_metadata__['parameters']
+            param_dict = dict(zip(target_params, args))
+
+            # This is necessary otherwise in some case TypeVar may be same:
+            #
+            # class A(BaseModel, Generic[T]): ...
+            # class B(A[int], Generic[T]): ...
+            # class C(B[str], Generic[T]): ...
+            #
+            key = '__pydantic_inserted_mro_origins__'
```
Comment (dmontagu) [inline pydantic/_internal/_model_construction.py:272]: This `__pydantic_inserted_mro_origins__` should be added as an annotated attribute on the `BaseModel` class if we are going to use it like this. I think it would also be reasonable to add it as a field in the `__pydantic_generic_metadata__` if that makes sense, but if we are using a new attribute in `BaseModel` it should be added near this area: https://github.com/pydantic/pydantic/blob/ae7770375837020fca5c182acddae8ac66302452/pydantic/main.py#L141

**Review by dmontagu — CHANGES_REQUESTED** [review]
See https://github.com/pydantic/pydantic/pull/10100/files#r1721798498; once that is addressed I'm good with this PR.

**Comment by kc0506** [comment]
@dmontagu I found that my current solution cannot handle this (very unlikely) case:
```py
class A(BaseModel, Generic[T1]): ...
class B(A[T2], Generic[T2]): ...
class C(B[T3], Generic[T3]): ...
class D(C[T1], Generic[T1]): ...
class E(D[T2], Generic[T2]):...
```
because the MRO of `D[T2]` will be like
```
D[~T2], D,
C[~T2], C[~T1], C,
B[~T2], B[~T1], B[~T3], B,
A[~T2], A[~T1], A[~T3], A[~T2], A, ...
```
and the repeated `A[T2]` is invalid.

Do you think this is worth fixing, or could we mark it as an expected fail and address it in other PR?

**Comment by dmontagu** [comment]
I'd say it's at least worth adding an xfailing test

**Comment by sydney-runkle** [comment]
@kc0506, if you're able to finish this up in the next couple of days, we can get this into our v2.9 release! Thanks for your awesome work here.

**Comment by kc0506** [comment]
Sure, I will try to wrap it up by this weekend.

**Comment by kc0506** [comment]
I've updated the logic insdie `mro`. The new approach does not need an extra class variable and can solve the failed case mentioned above, as well as remove redundant bases (e.g. `A[int], A[T2], A[T1], A, ...` will now become simply `A[int], A, ...`).

I think this is ready to be merged. Looking forward to any feedbacks!

**Review by dmontagu — APPROVED** [review]
I won't be surprised if this still ends up needing to be tweaked in some way in the future, but this is closer to what I had imagined the relevant code should look like, great improvement.

LGTM

**Comment by sydney-runkle** [comment]
@kc0506,

Awesome work here. I'm going to do another review here as well, then we'll merge this after our v2.9 release (we've already released the beta, so can't introduce any major changes during this beta trial period).

**Review by sydney-runkle — APPROVED** [review]
Amazing work here. Thanks a bunch for your contribution!!

I'm a bit worried about a few of the unchecked dict accesses, so let's make sure no unexpected behavior is witnessed here leading up to our v2.10 release.
